chain/jsonrpc/src/lib.rs

```diff
@@ -283,16 +287,68 @@ impl JsonRpcHandler {
         .map_err(|_| timeout_err())?
     }
 
+    async fn send_tx(
+        &self,
+        tx: SignedTransaction,
+        check_only: bool,
+    ) -> Result<NetworkClientResponses, RpcError> {
+        Ok(self
+            .client_addr
+            .send(NetworkClientMessages::Transaction {
+                transaction: tx,
+                is_forwarded: false,
+                check_only,
+            })
+            .map_err(|err| RpcError::server_error(Some(ServerError::from(err))))
+            .await?)
+    }
+
+    async fn send_tx_sync(&self, params: Option<Value>) -> Result<Value, RpcError> {
+        self.send_or_check_tx(params, false).await
+    }
+
+    async fn check_tx(&self, params: Option<Value>) -> Result<Value, RpcError> {
+        self.send_or_check_tx(params, true).await
+    }
+
+    async fn send_or_check_tx(
+        &self,
+        params: Option<Value>,
+        check_only: bool,
+    ) -> Result<Value, RpcError> {
+        let tx = parse_tx(params)?;
+        let tx_hash = (&tx.get_hash()).to_base();
+        match self.send_tx(tx, check_only).await? {
+            NetworkClientResponses::ValidTx => {
+                if check_only {
+                    Ok(Value::Null)
+                } else {
+                    Ok(Value::String(tx_hash))
+                }
+            }
+            NetworkClientResponses::RequestRouted => {
+                if check_only {
+                    Ok(Value::String("Node doesn't track this shard. Cannot determine whether the transaction is valid".to_string()))
+                } else {
+                    Ok(Value::String("Transaction is routed".to_string()))
+                }
+            }
```

Comment:
1. The return value should be an object (JSON key-value). Consider the RPC client perspective, you will need to guess if the returned string an error or a transaction hash; that is going to feel unreliable. We should always return an object (a structure / JSON key-value) instead of a plain string, this way it is also going to help us to extend the API response without breaking changes in the future (if needed).
2. This whole `check_only` looks a bit ugly to me, but I don't have better suggestions at the moment.

Reply:
>  you will need to guess if the returned string an error or a transaction hash; that is going to feel unreliable

You don't need to guess. They are two different rpcs

Reply:
> We should always return an object (a structure / JSON key-value) instead of a plain string

What do you suggest here?

Reply:
> They are two different rpcs

I was referring to the fact that the RPC can either return `"some-tx-hash"` or `"Node does't track this..."`, so from the type perspective they are both strings. To distinguish them you need to either hardcode the error text messages or try to parse that string as hash (check base58 validity).

Reply:
> They are two different rpcs

I was referring to the fact that the RPC can either return `"some-tx-hash"` or `"Node does't track this..."`, so from the type perspective they are both strings. To distinguish them you need to either hardcode the error text messages or try to parse that string as hash (check base58 validity).

Currently, the response is:

```js
{
  "jsonrpc": "2.0",
  "result": "some-tx-hash"
}
```

or 

```js
{
  "jsonrpc": "2.0",
  "result": "Node doesn't track ..."
}
```

This is really hard to work with, and also it is not extendible without a breaking change.

I suggest to have:

```js
{
  "jsonrpc": "2.0",
  "result": {
    "transaction_hash": "some-tx-hash",
  }
}
```

and

```js
{
  "jsonrpc": "2.0",
  "error": {
    "code": ...
    "data": "Node doesn't track...",
    "message": "Server error"
  }
}
```

Reply:
Well, I just noticed that the OK response is null value. Can we provide anything helpful in the response? (e.g. block height, block hash, chunk hash the transaction landed, if that is trivial to get)

Reply:
That is for checking the transaction. If you just want to check whether the transaction is valid I don't know what else you want to include in the response.